librosa/segment.py

```diff
@@ -1220,3 +1262,94 @@ def path_enhance(
         np.clip(R_smooth, 0, None, out=R_smooth)  # type: ignore
 
     return np.asanyarray(R_smooth)
+
+
+def __affinity_bandwidth(rec, bw_mode, k):
+    # rec should be a csr_matrix
+
+    # the api allows users to specify a scalar bandwidth directly, besides the string based options.
+    if isinstance(bw_mode, np.ndarray):
+        bandwidth = bw_mode
+        # check if bw is the right size
+        if bandwidth.shape != rec.shape:
+            raise ParameterError(
+                "Invalid matrix bandwidth with bad shape: {}."
+                "Should be {}.".format(bandwidth.shape, rec.shape)
+            )
+        if (bandwidth <= 0).any():
+            raise ParameterError(
+                "Invalid bandwidth. All entries must be strictly positive."
+            )
+        return bandwidth[rec.nonzero()]
+
+    elif isinstance(bw_mode, (int, float)):
+        bandwidth = bw_mode
+        if bandwidth <= 0:
+            raise ParameterError(
+                "Invalid scalar bandwidth={}. Must be strictly positive.".format(bandwidth)
+            )
+        return bandwidth
+
+    if bw_mode is None:
+        bw_mode = 'med_k_scalar'
+
+    if bw_mode not in ['med_k_scalar', 'mean_k', 'gmean_k', 'mean_k_avg', 'gmean_k_avg', 'mean_k_avg_and_pair']:
+        raise ParameterError(
+            (
+                "Invalid bandwidth='{}'. Must be either a positive scalar or one of "
+                "['med_k_scalar', 'mean_k', 'gmean_k', 'mean_k_avg', 'gmean_k_avg', 'mean_k_avg_and_pair']"
+            ).format(bw_mode)
+        )
+
+    # build a list of list that stores the distances to k nearest neighbors for all t points.
+    t = rec.shape[0]
+    knn_dists = []
+    for i in range(t):
+        # Get the links from point i
+        links = rec[i].nonzero()[1]
+        # catch empty dists lists in knn_dists
+        if len(links) == 0:
+            raise ParameterError("The sample at time point {} has no neighbors".format(i))
+
+        # Compute k nearest neighbors' distance and sort ascending
+        knn_dist_row = np.sort(rec[i, links].toarray()[0])[:k]
+        knn_dists.append(knn_dist_row)
+
+    # take the last element of each list for the distance to kth neighbor
+    
+    
+    dist_to_k = np.asarray([dists[-1] for dists in knn_dists])
+    avg_dist_to_first_ks = np.asarray([np.mean(dists) for dists in knn_dists])
```

Comment:
This seems like it could be faster by collecting knn_dists into an ndarray first, and then using `np.mean(knn_dists, axis=-1)`

Reply:
I think I did it this way because the lists `dists` in `knn_dists` can be of different length, and won't collect nicely into an ndarray... Not exactly sure when different lengths `dists` list would pop up, but I'll leave this potential efficiency issue for another PR if that's ok.